program/rust/src/rust_oracle.rs

```diff
@@ -87,6 +95,68 @@ pub fn init_mapping(
     Ok(SUCCESS)
 }
 
+/// add a price account to a product account
+/// accounts[0] funding account                                   [signer writable]
+/// accounts[1] product account to add the price account to       [signer writable]
+/// accounts[2] newly created price account                       [signer writable]
+pub fn add_price(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> OracleResult {
+    if instruction_data.len() < size_of::<cmd_add_price_t>() {
+        return Err(ProgramError::InvalidArgument);
+    }
+    let cmd_args = load::<cmd_add_price_t>(instruction_data)?;
+
+    if cmd_args.expo_ > PC_MAX_NUM_DECIMALS as i32
+        || cmd_args.expo_ < -(PC_MAX_NUM_DECIMALS as i32)
+        || cmd_args.ptype_ == PC_PTYPE_UNKNOWN
+    {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    let [_funding_account, product_account, price_account] = match accounts {
+        [x, y, z]
+            if valid_funding_account(x)
+                && valid_signable_account(program_id, y, PC_PROD_ACC_SIZE as usize)
+                && valid_signable_account(program_id, z, size_of::<pc_price_t>()) =>
+        {
+            Ok([x, y, z])
+        }
+        _ => Err(ProgramError::InvalidArgument),
+    }?;
+
+    let mut product_data = load_account_as_mut::<pc_prod_t>(product_account)?;
+    {
+        let price_data = load_account_as::<pc_price_t>(price_account)?;
+
+        if product_data.magic_ != PC_MAGIC
+            || product_data.ver_ != cmd_args.ver_
+            || product_data.type_ != PC_ACCTYPE_PRODUCT
+            || price_data.magic_ != 0
```

Comment:
One thing i've noticed from doing this translation exercise is that the check for whether an account is uninitialized differs in all of the instructions. `init_mapping` checks both the magic and ver fields, but `add_account` only checks the magic. I think it should be fine for us to standardize on one of these and encode that in `validate_fresh_account`.